libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtils.kt

```diff
@@ -57,47 +59,93 @@ object MapboxRouteLineUtils {
         expressionBuilder.lineProgress()
-        expressionBuilder.stop {
-            rgba {
-                literal(0.0)
-                literal(0.0)
-                literal(0.0)
-                literal(0.0)
+        expressionBuilder.color(lineStartColor)
+
+        getFilteredRouteLineExpressionData(
+            distanceOffset,
+            routeLineExpressionData,
+            lineColor
+        ).forEach {
+            // If the color hasn't changed there's no reason to add it to the expression. A smaller
+            // expression is less work for the map to process.
+            if (it.segmentColor != lastColor) {
+                lastColor = it.segmentColor
+                expressionBuilder.stop {
+                    literal(it.offset)
+                    color(it.segmentColor)
+                }
             }
         }
-        val filteredItems = routeLineExpressionData
-            .filter { it.offset > distanceOffset }.distinctBy { it.offset }
-        when (filteredItems.isEmpty()) {
-            true -> when (routeLineExpressionData.isEmpty()) {
-                true -> listOf(RouteLineExpressionData(distanceOffset, fallbackRouteColor))
-                false -> listOf(routeLineExpressionData.last().copy(offset = distanceOffset))
+        return expressionBuilder.build()
+    }
+
+    /**
+     * Returns an [Expression] for a gradient line that will start with the @param lineBaseColor,
+     * creating the first step at the @param distanceOffset with additional steps
+     * according to the items in the @param routeLineExpressionData.
+     *
+     * @param distanceOffset a distance value representing the first step in the [Expression]
+     * @param routeLineExpressionData a collection of [RouteLineExpressionData] items
+     * @param lineBaseColor the starting color for the line
+     * @param defaultColor the default color for the line
+     * @param substitutionColor the color to be used in lieu of the default color if active leg
+     * index is greater than or equal to 0 and the leg index of a given [RouteLineExpressionData]
+     * is not equal to the @param activeLegIndex.
+     *
+     * @return an [Expression] intended to be used as a gradient line
+     */
+    internal fun getRouteLineExpression(
+        distanceOffset: Double,
+        routeLineExpressionData: List<RouteLineExpressionData>,
+        lineBaseColor: Int,
+        defaultColor: Int,
+        substitutionColor: Int,
+        activeLegIndex: Int
+    ): Expression {
+        var lastColor = Int.MAX_VALUE
+        val expressionBuilder = Expression.ExpressionBuilder("step")
+        expressionBuilder.lineProgress()
+        expressionBuilder.color(lineBaseColor)
+
+        getFilteredRouteLineExpressionData(
+            distanceOffset,
+            routeLineExpressionData,
+            defaultColor
+        ).forEach {
+            val colorToUse = if (activeLegIndex >= 0 && it.legIndex != activeLegIndex) {
```

Comment:
nit: `activeLegIndex >= 0` is an assumption based on `MapboxRouteLineApi#INVALID_ACTIVE_LEG_INDEX` that is decoupled from this processor. Someone interacting with this method without the context could misuse it. Let's document this behavior or make it type safe, for example: `activeLegIndex: Int?`.

Reply:
It seems less type safe to me to make `activeLegIndex` nullable since a leg index less than zero is invalid.  To me the implementation is protecting itself. In addition this is an internal method so external developers won't be calling it.

Reply:
Sounds good, no strong opinion. Would just be good to add a doc entry for our future selves.